processor/marshalprocessor/processor.go

```diff
@@ -0,0 +1,136 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package marshalprocessor
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.uber.org/zap"
+)
+
+type marshalProcessor struct {
+	logger          *zap.Logger
+	marshalTo       string
+	kvSeparator     rune
+	kvPairSeparator rune
+}
+
+func newMarshalProcessor(logger *zap.Logger, cfg *Config) *marshalProcessor {
+	return &marshalProcessor{
+		logger:          logger,
+		marshalTo:       cfg.MarshalTo,
+		kvSeparator:     cfg.KVSeparator,
+		kvPairSeparator: cfg.KVPairSeparator,
+	}
+}
+
+func (mp *marshalProcessor) processLogs(_ context.Context, ld plog.Logs) (plog.Logs, error) {
+	for i := 0; i < ld.ResourceLogs().Len(); i++ {
+		resourceLog := ld.ResourceLogs().At(i)
+		for j := 0; j < resourceLog.ScopeLogs().Len(); j++ {
+			scopeLog := resourceLog.ScopeLogs().At(j)
+			for k := 0; k < scopeLog.LogRecords().Len(); k++ {
+				logRecord := scopeLog.LogRecords().At(k)
+				logBody := logRecord.Body()
+				// If body is not a map, skip that log
+				if logBody.Type() != pcommon.ValueTypeMap {
+					mp.logger.Warn("Non map body not supported", zap.Any("body", logBody))
+					continue
+				}
+				switch strings.ToUpper(mp.marshalTo) {
+				case "JSON":
+					jsonBody := logBody.AsString()
+					logBody.SetStr(jsonBody)
+				case "XML":
+					return ld, fmt.Errorf("XML not yet supported")
+				case "KV":
+					kvBody := mp.convertToKV(logBody.Map())
+					logBody.SetStr(kvBody)
+				default:
+					return ld, fmt.Errorf("Unrecognized format to marshal to: %s", mp.marshalTo)
+				}
+			}
+		}
+	}
+
+	return ld, nil
+}
+
+func (mp *marshalProcessor) convertToKV(logBody pcommon.Map) string {
+	var kvStrings []string
+	var kvPairSeparator string
+	var kvSeparator string
+	if mp.kvPairSeparator == 0 {
+		kvPairSeparator = " "
+	} else {
+		kvPairSeparator = string(mp.kvPairSeparator)
+	}
+	if mp.kvSeparator == 0 {
+		kvSeparator = "="
+	} else {
+		kvSeparator = string(mp.kvSeparator)
+	}
+
+	for k, v := range logBody.AsRaw() {
+		k = strings.ReplaceAll(k, "\"", "\\\"")
+		if strings.Contains(k, kvPairSeparator) || strings.Contains(k, kvSeparator) {
+			k = "\"" + k + "\""
+		}
+
+		if valMap, ok := v.(map[string]interface{}); ok {
+			v = convertMapToString(valMap)
+		}
+
+		v = strings.ReplaceAll(fmt.Sprintf("%v", v), "\"", "\\\"")
+		if strings.Contains(fmt.Sprintf("%v", v), kvPairSeparator) || strings.Contains(fmt.Sprintf("%v", v), kvSeparator) {
+			v = "\"" + fmt.Sprintf("%v", v) + "\""
+		}
+
+		kvStrings = append(kvStrings, fmt.Sprintf("%s%s%v", k, kvSeparator, v))
+	}
+
+	sort.Strings(kvStrings)
```

Comment:
I'm not sure if this will necessarily sort by the keys in all cases. Could we do something similar to what we do in convertMapToString instead, sorting explicitly on the keys?

Reply:
@BinaryFissionGames I think this will properly sort by keys, when might it not? The unit tests are consistently passing with deeply nested values being sorted in the same order each run.

Reply:
I guess the quoting of keys will change the order, and rare cases where the separator is found in the key.